File: 0034-find-first-and-last-position-of-element-in-sorted-array/0034-find-first-and-last-position-of-element-in-sorted-array.cpp
class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {

        //optimal solution -> O(logn) in java

        if(nums.empty()){

            return {-1,-1};
        }

        vector<int> res(2,-1);

        int i = 0;
        while(i < nums.size()){

            if(target != nums[i]){

               i++;
            }else{

                break;
            }
        }

        if(i == nums.size()){

            return {-1,-1};
        }

        int j = i;

        res[0] = i;

        while(j < nums.size() && nums[j] == target){

            j++;
        }

        res[1] = j-1;

        return res;
    }
};

File: 0034-find-first-and-last-position-of-element-in-sorted-array/0034-find-first-and-last-position-of-element-in-sorted-array.java
class Solution {

    public int[] searchRange(int[] nums, int target) {
        

        int left = findLeftBound(nums, target);
        int right = findRightBound(nums, target);

        return new int[]{left, right};
    }

    private int findLeftBound(int[] nums, int target){

        int idx = -1, low = 0, high = nums.length - 1;

        while(low <= high){

            int mid = low + (high - low) / 2;

            if(nums[mid] == target){

                idx = mid;
                high = mid - 1;
            } else if(nums[mid] < target){

                low = mid + 1;
            } else{

                high = mid - 1;
            }
        }

        return idx;
    }

    private int findRightBound(int[] nums, int target){

        int idx = -1, low = 0, high = nums.length - 1;

        while(low <= high){

            int mid = low + (high - low) / 2;

            if(nums[mid] == target){

                idx = mid;
                low = mid + 1;
            } else if(nums[mid] < target){

                low = mid + 1;
            } else{

                high = mid - 1;
            }
        }

        return idx;
    }

}